OSDOCS-1709 installation decision tree (#28085)
openshift/openshift-docs: installing/installing-preparing.adoc

[id="installing-preparing"]

> = Selecting a cluster installation method and preparing it for users

Reviewer: You could shorten this title. At present it seems crowded in the TOC on the left of the page on docs.openshift.com. You could shorten it to something like the following examples:

"Selecting an installation method"
"Selecting and preparing an installation method"

Author: I don't think that either of those accurately reflects the content, though. The prep steps are for the space between installation and productive use of the cluster.

toc::[]

Before you install {product-title}, think about the following questions to decide what kind of installation process to follow and make sure you that you have all of the resources that you need to prepare your cluster for your users.

[id="installing-preparing-selecting-cluster-type"]
== Selecting your cluster installation type
Before you install a {product-title} cluster, you need to select the best installation instructions to follow. Think about your answers to the following questions to select the best option.

[id="installing-preparing-install-manage"]
=== Do you want to install and manage an {product-title} cluster yourself?

If you want to install and manage {product-title} yourself, you can install it on the following platforms:

* Amazon Web Services (AWS)
* Microsoft Azure
* Google Cloud Platform (GCP)
* {rh-openstack}
* {rh-virtualization}
* IBM Z and LinuxONE
* IBM Power
* VMware vSphere
* Bare metal or other platform agnostic infrastructure

You can deploy an {product-title} 4 cluster to both on-premise data centers and to cloud hosting services, but all of the machines in your cluster must be in the same datacenter or cloud hosting service.

> 

Reviewer: This is where the ".Deploying {product-title} by using a managed service" title that I suggested in my previous comment could go.

If you want to use {product-title} but do not want to manage your cluster yourself, you have several managed service options. If you want a cluster that is fully managed by Red Hat, you can use link:https://www.openshift.com/products/dedicated/[OpenShift Dedicated] or link:https://www.openshift.com/products/online/[OpenShift Online]. You can also use OpenShift as a managed service on Azure, AWS, IBM Cloud, or Google Cloud. For more information about managed services, see the link:https://www.openshift.com/products[OpenShift Products] page.

[id="installing-preparing-migrate"]
=== Have you used {product-title} 3 and want to use {product-title} 4?

> If you used {product-title} 3 and want to try {product-title} 4, you need to understand how different {product-title} 4 is. {product-title} 4 weaves the Operators that comprise the program and the operating system that the platform runs on, {op-system-first}, together seamlessly. Instead of deploying machines and configuring their operating systems so that you can install {product-title} on them, the {op-system} operating system is an integral part of the {product-title} cluster. Your cluster deploys the operating system for the cluster machines as part of the program itself. The easiest way to use {product-title} is by letting it make all of the infrastructure and machines that it needs.

Reviewer 2: Boo to that. :(

Because you need to provision machines as part of the {product-title} cluster installation process, you cannot upgrade an {product-title} 3 cluster to {product-title} 4. Instead, you must create a new {product-title} 4 cluster and migrate your {product-title} 3 workloads to them. For more information about migrating, see link:https://redhat-cop.github.io/openshift-migration-best-practices/01-planning.html[OpenShift Migration Best Practices]. Because you must migrate to {product-title} 4, you can use any type of production cluster installation process to create your new cluster.

[id="installing-preparing-existing-components"]

> === How many existing components do you want to use with your cluster?

Reviewer: I would say that the top level question for this section is whether a reader wants to use existing components at all, as opposed to how many they need to use. Depending on whether you agree with this, you could change the title to something like the following example:

` === Do you want to use existing components in your cluster?`

> 

Reviewer: I think this section would benefit from some dot titles to help readers scan the content more easily. For example you could include two dot titles as follows:

```
.Customizing installer-provisioned infrastructure installations
.Installing {product-title} by using user-provisioned infrastructure
```
The first of those could go here. I have added another comment where the second title could potentially go.

Because of the way that the operating system and platform intertwine, it is far easier to complete an installation where you let the installation program for {product-title} stand up all of the infrastructure. These are called _installer provisioned infrastructure_ installations. In this type of installation, you can provide some existing infrastructure to the cluster, but the installation program deploys all of the machines that your cluster initially needs.

You can deploy an installer-provisioned infrastructure cluster without specifying any customization to the cluster or its underlying machines to xref:../installing/installing_aws/installing-aws-default.adoc#installing-aws-default[AWS], xref:../installing/installing_azure/installing-azure-default.adoc#installing-azure-default[Azure], or xref:../installing/installing_gcp/installing-gcp-default.adoc#installing-gcp-default[GCP]. These installation methods are the fastest way to deploy a production-capable {product-title} cluster.

> If you need to perform basic configuration for your installer-provisioned infrastructure cluster, such as the region in your cloud that you deploy the machines to and the instance type for the cluster machines, you can customize an installation for xref:../installing/installing_aws/installing-aws-customizations.adoc#installing-aws-customizations[AWS], xref:../installing/installing_azure/installing-azure-customizations.adoc#installing-azure-customizations[Azure], or xref:../installing/installing_gcp/installing-gcp-customizations.adoc#installing-gcp-customizations[GCP].

Reviewer: To me it is not obvious that there is a difference between each of the pages that are linked to "AWS", "Azure" and "GCP" throughout the assembly, without opening them and reviewing each. It might be clearer if you provide a bulleted list that includes the full assembly title for each respective document. Here is an example for this line:

```
If you need to perform basic configuration for your installer-provisioned infrastructure cluster, such as the region in your cloud that you deploy the machines to and the instance type for the cluster machines, you can customize an installation. See the following documents for platform-specific information about customizing an installation:

* xref:../installing/installing_aws/installing-aws-customizations.adoc#installing-aws-customizations[Installing a cluster on AWS with customizations], 
* xref:../installing/installing_azure/installing-azure-customizations.adoc#installing-azure-customizations[Installing a cluster on Azure with customizations]
* xref:../installing/installing_gcp/installing-gcp-customizations.adoc#installing-gcp-customizations[Installing a cluster on GCP with customizations].
```
The same applies throughout the document where we are pointing the reader to specific sections of the installation pages. It might be useful if we state the name of the assembly each time, so that it is easy to distinguish between each suggested link. I won't highlight each one in this review.

> If you need to perform basic configuration for your installer-provisioned infrastructure cluster, such as the region in your cloud that you deploy the machines to and the instance type for the cluster machines, you can customize an installation for xref:../installing/installing_aws/installing-aws-customizations.adoc#installing-aws-customizations[AWS], xref:../installing/installing_azure/installing-azure-customizations.adoc#installing-azure-customizations[Azure], or xref:../installing/installing_gcp/installing-gcp-customizations.adoc#installing-gcp-customizations[GCP].

Reviewer 2: I would not call out the region as a configuration item here that the user **might** need to configure. The user must always configure the region.

> For installer-provisioned infrastructure installations, you can use an existing xref:../installing/installing_aws/installing-aws-vpc.adoc#installing-aws-vpc[VPC in AWS], xref:../installing/installing_azure/installing-azure-vnet.adoc#installing-azure-vnet[vNet in Azure], or xref:../installing/installing_gcp/installing-gcp-vpc.adoc#installing-gcp-vpc[VPC in GCP]. You can also reuse part of your networking so that your cluster in xref:../installing/installing_aws/installing-aws-network-customizations.adoc#installing-aws-network-customizations[AWS], xref:../installing/installing_azure/installing-azure-network-customizations.adoc#installing-azure-network-customizations[Azure], or xref:../installing/installing_gcp/installing-gcp-network-customizations.adoc#installing-gcp-network-customizations[GCP] can coexist with existing IP address allocations in your environment and integrate with existing MTU and VXLAN configurations. If you have existing accounts and credentials on these clouds, you can modify the accounts to have the required permissions to install {product-title} clusters on them.

Reviewer: Should "part of your networking" be "part of your networking infrastructure" or something like that?

> For installer-provisioned infrastructure installations, you can use an existing xref:../installing/installing_aws/installing-aws-vpc.adoc#installing-aws-vpc[VPC in AWS], xref:../installing/installing_azure/installing-azure-vnet.adoc#installing-azure-vnet[vNet in Azure], or xref:../installing/installing_gcp/installing-gcp-vpc.adoc#installing-gcp-vpc[VPC in GCP]. You can also reuse part of your networking so that your cluster in xref:../installing/installing_aws/installing-aws-network-customizations.adoc#installing-aws-network-customizations[AWS], xref:../installing/installing_azure/installing-azure-network-customizations.adoc#installing-azure-network-customizations[Azure], or xref:../installing/installing_gcp/installing-gcp-network-customizations.adoc#installing-gcp-network-customizations[GCP] can coexist with existing IP address allocations in your environment and integrate with existing MTU and VXLAN configurations. If you have existing accounts and credentials on these clouds, you can modify the accounts to have the required permissions to install {product-title} clusters on them.

Reviewer 2: ```
If you have existing accounts and credentials on these clouds, you can modify the accounts to have the required permissions to install {product-title} clusters on them.
```
This sentence does not seem to fit into this section. This section is about custom configurations that you user can do while still doing an installer-provisioned installation. This sentence is more about what the user must do to configure their account. Regardless of whether the user has an existing account or is creating a new account, the user still needs to configure the required permissions for the account. Maybe what you want to say is something like the following.
```
If you have existing accounts and credentials on these clouds, you can re-use them, but you may need to modify the accounts to have the required permissions to install {product-title} clusters on them.
```

> You can use the installer-provisioned infrastructure method to create appropriate machine instance in your data center for xref:../installing/installing_rhv/installing-rhv-default.adoc#installing-rhv-default[{rh-virtualization}], xref:../installing/installing_vsphere/installing-vsphere-installer-provisioned.adoc#installing-vsphere-installer-provisioned[vSphere], and xref:../installing/installing_bare_metal_ipi/ipi-install-overview#ipi-install-overview[bare metal].

Reviewer: You could change "to create appropriate machine instance" to "to create the appropriate machine instances".

> 

Reviewer: As a follow up to my earlier comment, this is where you could add the following dot title to help the reader scan this content:

`.Installing {product-title} by using user-provisioned infrastructure`

> If you want to reuse extensive cloud infrastructure, you can complete a _user provisioned infrastructure_ installation. With these installations, you manually deploy the machines that your cluster requires during the installation process. If you perform a user provisioned infrastructure installation on xref:../installing/installing_aws/installing-aws-user-infra.adoc#installing-aws-user-infra[AWS], xref:../installing/installing_azure/installing-azure-user-infra.adoc#installing-azure-user-infra[Azure], or xref:../installing/installing_gcp/installing-gcp-user-infra.adoc#installing-gcp-user-infra[GCP], you can use the provided templates to help you stand up all of the required components. You can also use the provider agnostic installation method to deploy a cluster into other clouds.

Reviewer: Should "provider agnostic" be "provider-agnostic"?

Reviewer: Should "user provisioned infrastructure" be "user-provisioned infrastructure". This will add consistency with the references to "installer-provisioned infrastructure" in this assembly. This occurs in other places in this PR also, but I won't highlight them all.

> You can also complete a user provisioned infrastructure installation on hardware in your data center. If you use xref:../installing/installing_openstack/installing-openstack-installer-custom.adoc#installing-openstack-installer-custom[{rh-openstack}], xref:../installing/installing_rhv/installing-rhv-user-infra.adoc#installing-rhv-user-infra[{rh-virtualization}], xref:../installing/installing_ibm_z/installing-ibm-z.adoc#installing-ibm-z[IBM Z or LinuxONE], xref:../installing/installing_ibm_power/installing-ibm-power.adoc#installing-ibm-power[IBM Power], or xref:../installing/installing_vsphere/installing-vsphere.adoc#installing-vsphere[vSphere], use the specific installation instructions to deploy your cluster. If you use other supported hardware, follow the xref:../installing/installing_bare_metal/installing-bare-metal.adoc#installing-bare-metal[bare metal installation] procedure.

Reviewer: You could change "on hardware in your data center" to "on your own hardware", because the reader's hardware might not be in a data centre.

> You can also complete a user provisioned infrastructure installation on hardware in your data center. If you use xref:../installing/installing_openstack/installing-openstack-installer-custom.adoc#installing-openstack-installer-custom[{rh-openstack}], xref:../installing/installing_rhv/installing-rhv-user-infra.adoc#installing-rhv-user-infra[{rh-virtualization}], xref:../installing/installing_ibm_z/installing-ibm-z.adoc#installing-ibm-z[IBM Z or LinuxONE], xref:../installing/installing_ibm_power/installing-ibm-power.adoc#installing-ibm-power[IBM Power], or xref:../installing/installing_vsphere/installing-vsphere.adoc#installing-vsphere[vSphere], use the specific installation instructions to deploy your cluster. If you use other supported hardware, follow the xref:../installing/installing_bare_metal/installing-bare-metal.adoc#installing-bare-metal[bare metal installation] procedure.

Reviewer 2: nit: Can we find another way of phrasing "You can also"? I find it a bit jarring to see two consecutive sentences start that same way.

[id="installing-preparing-security"]

> === Do you need extra security for your cluster?

Reviewer 2: I would order this section in somewhat increasing levels of security. When I read it the first time, I got to the first section where it talks about "limited access to the internet" and conflated that with a private cluster. Then I was confused as to why that required user-provisioned installation.
1. Proxy
2. Private
3. Disconnected/Restricted
4. Gov cloud
5. FIPS

> If you need to install your cluster so that it has limited access to the internet, such as a disconnected or restricted network cluster, you can xref:../installing/install_config/installing-restricted-networks-preparations.adoc#installing-restricted-networks-preparations[mirror the installation packages] and install the cluster from them. Follow detailed instructions for user provisioned infrastructure installations into restricted networks for xref:../installing/installing_aws/installing-restricted-networks-aws.adoc#installing-restricted-networks-aws[AWS], xref:../installing/installing_gcp/installing-restricted-networks-gcp.adoc#installing-restricted-networks-gcp[GCP], xref:../installing/installing_ibm_z/installing-restricted-networks-ibm-z.adoc#installing-restricted-networks-ibm-z[IBM Z or LinuxONE], xref:../installing/installing_ibm_power/installing-restricted-networks-ibm-power.adoc#installing-restricted-networks-ibm-power[IBM Power], xref:../installing/installing_vsphere/installing-restricted-networks-vsphere.adoc#installing-restricted-networks-vsphere[vSphere], or xref:../installing/installing_bare_metal/installing-restricted-networks-bare-metal.adoc#installing-restricted-networks-bare-metal[bare metal].

Reviewer: You could change the first sentence to the following:

"If you need to install a cluster that has limited access to the internet, such as a disconnected or restricted network cluster, you can..."

> If you need to install your cluster so that it has limited access to the internet, such as a disconnected or restricted network cluster, you can xref:../installing/install_config/installing-restricted-networks-preparations.adoc#installing-restricted-networks-preparations[mirror the installation packages] and install the cluster from them. Follow detailed instructions for user provisioned infrastructure installations into restricted networks for xref:../installing/installing_aws/installing-restricted-networks-aws.adoc#installing-restricted-networks-aws[AWS], xref:../installing/installing_gcp/installing-restricted-networks-gcp.adoc#installing-restricted-networks-gcp[GCP], xref:../installing/installing_ibm_z/installing-restricted-networks-ibm-z.adoc#installing-restricted-networks-ibm-z[IBM Z or LinuxONE], xref:../installing/installing_ibm_power/installing-restricted-networks-ibm-power.adoc#installing-restricted-networks-ibm-power[IBM Power], xref:../installing/installing_vsphere/installing-restricted-networks-vsphere.adoc#installing-restricted-networks-vsphere[vSphere], or xref:../installing/installing_bare_metal/installing-restricted-networks-bare-metal.adoc#installing-restricted-networks-bare-metal[bare metal].

Reviewer 2: When talking about the public internet, capitalize it--Internet.

If you use a user-provisioned installation method, you can configure a proxy for your cluster. The instructions are included in each installation procedure.

If you want to prevent your cluster on a public cloud from exposing endpoints externally, you can deploy a private cluster with installer-provisioned infrastructure on xref:../installing/installing_aws/installing-aws-private.adoc#installing-aws-private[AWS], xref:../installing/installing_azure/installing-azure-private.adoc#installing-azure-private[Azure], or xref:../installing/installing_gcp/installing-gcp-private.adoc#installing-gcp-private[GCP].

You can also configure the cluster machines to use xref:../installing/installing-fips.adoc#installing-fips[FIPS Validated / Modules in Process cryptographic libraries] during installation.

[id="installing-preparing-cluster-for-users"]

> == Preparing your cluster for users

Reviewer: For findability and to help convey the topic in more detail, you could add to this title as follows:

`== Preparing your cluster for users after an installation`

> Because you do not have the opportunity to configure all of the supporting infrastructure for your cluster during installation, you need to perform additional configuration between the completion of cluster deployment and when your users access the cluster. You can xref:../installing/install_config/customizations.adoc#customizations[customize] the Operators that make up your cluster and integrate other required systems, such as an identity provider.

Reviewer: It might be useful to readers who are not familiar with Operators to highlight why you would want to customise the Operators (i.e. that is how you customise OCP).

> Because you do not have the opportunity to configure all of the supporting infrastructure for your cluster during installation, you need to perform additional configuration between the completion of cluster deployment and when your users access the cluster. You can xref:../installing/install_config/customizations.adoc#customizations[customize] the Operators that make up your cluster and integrate other required systems, such as an identity provider.

Reviewer 2: Starting with `Because` gives me the impression that this is a deficiency of OpenShift.
```suggestion
Some configuration is not required to install the cluster but should be done before your users access the cluster. You can xref:../installing/install_config/customizations.adoc#customizations[customize] the Operators that make up your cluster and integrate other required systems, such as an identity provider.
```

//This link will change when we consolidate the customizations page with the post-installation activities.

For a production cluster, you must configure the following integrations:

* xref:../storage/understanding-persistent-storage.adoc#understanding-persistent-storage[Persistent storage]
* xref:../authentication/understanding-identity-provider.adoc#understanding-identity-provider[An identity provider]

> * xref:../monitoring/configuring-the-monitoring-stack.adoc#configuring-the-monitoring-stack[Monitoring] for the cluster itself

Reviewer: In line with the terminology used in the Monitoring book from 4.6 GA, I recommend changing this bullet point to the following:

* xref:../monitoring/configuring-the-monitoring-stack.adoc#configuring-the-monitoring-stack[Monitoring core OpenShift Container Platform components]

> * xref:../monitoring/enabling-monitoring-for-user-defined-projects.adoc#enabling-monitoring-for-user-defined-projects[Monitoring] for application workloads

Reviewer: I think that monitoring for user-defined projects is optional. The feature is not enabled by default and my understanding is that production clusters could feasibly run without this capability being enabled. I would recommend removing this bullet point from this list.

[id="installing-preparing-cluster-for-workloads"]
== Preparing your cluster for workloads

> Depending on your workload needs, you might need to take extra steps before you begin deploying applications. After you prepare infrastructure to support your application xref:../builds/build-strategies.adoc#build-strategies[build strategy], you might need to make provisions for xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#cnf-performance-addon-operator-for-low-latency-nodes[low-latency] workloads or xref:../nodes/pods/nodes-pods-secrets.adoc#nodes-pods-secrets[protect sensitive workloads].

Reviewer: You could add "For example," to the start of the second sentence, to show that it is linked to the first statement and to indicate that there might be other preparations in addition to these. I.e.:

"For example, after you prepare infrastructure to support your application..."

> Depending on your workload needs, you might need to take extra steps before you begin deploying applications. After you prepare infrastructure to support your application xref:../builds/build-strategies.adoc#build-strategies[build strategy], you might need to make provisions for xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#cnf-performance-addon-operator-for-low-latency-nodes[low-latency] workloads or xref:../nodes/pods/nodes-pods-secrets.adoc#nodes-pods-secrets[protect sensitive workloads].

Reviewer 2: ```suggestion
Depending on your workload needs, you might need to take extra steps before you begin deploying applications. After you prepare infrastructure to support your application xref:../builds/build-strategies.adoc#build-strategies[build strategy], you might need to make provisions for xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#cnf-performance-addon-operator-for-low-latency-nodes[low-latency] workloads or to xref:../nodes/pods/nodes-pods-secrets.adoc#nodes-pods-secrets[protect sensitive workloads].
```
